cpp/src/lance/format/schema.cc

```diff
@@ -224,7 +223,7 @@ std::shared_ptr<lance::encodings::Encoder> Field::GetEncoder(
 ::arrow::Result<std::shared_ptr<lance::encodings::Decoder>> Field::GetDecoder(
     std::shared_ptr<::arrow::io::RandomAccessFile> infile) {
   std::shared_ptr<lance::encodings::Decoder> decoder;
-  auto data_type = type();
+  auto data_type = storage_type();
```

Comment:
not sure where this should go in code comments / docs but let's formalize the abstraction that the encoder/decoder only deals with storage types when reading/writing data.

Reply:
Added a comment to `Decoder` base class to mention that it does not accept ExtensionType